README.md
#  passwordfield: [^0.2.0](https://pub.dev/packages/passwordfield) 

<a href="https://pub.dev/packages/passwordfield"><img src="https://img.shields.io/pub/v/passwordfield.svg" alt="Pub"></a>

A simple and easy to use flutter package to add a passwordfield to your Flutter project.

## Getting Started

### Installation

- Add the dependency
```
flutter pub add passwordfield
```
- Import the package

```
import 'package:passwordfield/passwordfield.dart';
```
- Voila! use the Widget
```
PasswordField();
```

## Now Supports adding constraints on the Password 

![Screenshot from 2020-01-11 22-16-13](https://user-images.githubusercontent.com/31410839/72208800-2ca25e80-34cd-11ea-9cc2-8dd1274ff975.png)



<table>
<tr>
<th>Example 1</th>
<th>Output</th>
</tr>
<tr>
<td>

```dart
PasswordField(
  color: Colors.blue,
  passwordConstraint: r'.*[@$#.*].*',
  inputDecoration: PasswordDecoration(),
  hintText: 'must have special characters',
  border: PasswordBorder(
    border: OutlineInputBorder(
      borderSide: BorderSide(
        color: Colors.blue.shade100,
      ),
      borderRadius: BorderRadius.circular(12),
    ),
    focusedBorder: OutlineInputBorder(
      borderSide: BorderSide(
        color: Colors.blue.shade100,
      ),
      borderRadius: BorderRadius.circular(12),
    ),
    focusedErrorBorder: OutlineInputBorder(
      borderRadius: BorderRadius.circular(12),
      borderSide:
          BorderSide(width: 2, color: Colors.red.shade200),
    ),
  ),
  errorMessage:
      'must contain special character either . * @ # \$',
),
```

</td>
<td>

![Screen-Recording-2021-09-18-at-9](https://user-images.githubusercontent.com/31410839/133895843-3c8044d7-6829-4618-b983-305668b19d79.gif)


</td>
</tr>
</table>



<table>
<tr>
<th>Example 2</th>
<th>Output</th>
</tr>
<tr>
<td>      

```dart
PasswordField(
  backgroundColor: Colors.blue.withOpacity(0.2),
  errorMessage: '''
- A uppercase letter
- A lowercase letter
- A digit
- A special character
- A minimum length of 8 characters
 ''',
  hintText: 'Default password constraint ',
  inputDecoration: PasswordDecoration(
    inputPadding: const EdgeInsets.symmetric(horizontal: 20),
  ),
  border: PasswordBorder(
    border: OutlineInputBorder(
        borderSide:
            const BorderSide(width: 0, color: Colors.grey),
        borderRadius: BorderRadius.circular(25.7)),
    focusedBorder: OutlineInputBorder(
      borderSide:
          const BorderSide(width: 0, color: Colors.grey),
      borderRadius: BorderRadius.circular(25.7),
    ),
    enabledBorder: OutlineInputBorder(
      borderSide:
          const BorderSide(width: 0, color: Colors.grey),
      borderRadius: BorderRadius.circular(25.7),
    ),
  ),
),
```
</td>
<td>

![Screen-Recording-2021-09-18-at-1](https://user-images.githubusercontent.com/31410839/133895898-76fc11b7-3a6b-45dc-bbe6-637e971471f1.gif)

</td>
</tr>
</table>

### [More examples here 👈](example/lib/main.dart)

## Properties

```bool autoFocus```: if autofocus is true keyboard pops up as soon as the widget is rendered on screen defaults to false

```Color backgroundColor``` Background Color for the textfield (defaults to white), By default the backgroundColor is based on the current [Theme].

```PasswordBorder border```:  Border class for the passwordfield which provides borders for different states of the passwordfield e.g [enabledBorder],[border], [focusedBorder], [focusedErrorBorder]

```TextEditingController controller```: A controller for an editable passwordfield.

```int errorMaxLines```: Defaults to null, which means that the [errorText] will be limited  to a single line with [TextOverflow.ellipsis]

```String errorMessage```: custom message to show if the input password does not match the pattern.

```String floatingText```: floatingText to show when floatingPlaceholder is true

```hasFloatingPlaceholder```: whether the placeholder can float to left top on focus, defaults to false

```final String hintText```:  default text to show on the passwordfield when it is empty

```PasswordDecoration inputDecoration```: Decoration class for the PasswordField to customize the input styling

```int maxLength```: the max number of characters the password field can support
  
```Function onSubmit```: function triggerred when the submit button on keyboard is pressed

```Function onChanged```: function triggerred when the password is changed in the passwordfield

```String passwordConstraint```: supports adding constraints to the Passwordfield by adding a regex pattern
  
You are welcome to contribute to this package, to contribute please read the [contributing guidelines](CONTRIBUTING.md).


CHANGELOG.md
## [0.2.0]- 12 June 2023

- Migrate to flutter 3.0
- renamed `inputDecoration` to `passwordDecoration`

## [0.1.0]- 18 September 2021

- Migrate to [sound null-safety](https://dart.dev/null-safety)
- [Breaking changes]
  - separated the decoration and border apis, please check the readme for updated api changes
  - renamed `pattern` to `passwordConstraint`
- `onChanged` callback was added back

## [0.0.82]- 16 August 2020

- add floatingText label for floating placeholder
- improved documentation

## [0.0.81]- 09 August 2020

- minor fixes [Issue:Fix](https://github.com/maheshmnj/passwordfield-flutter-package/issues/3)

## [0.0.7]- 03 July 2020

- Added ability to add multiline error Message and style to errorMessage with proprty called errorMaxLines and errorStyle. [Issue:Feature Request](https://github.com/maheshmnj/passwordfield-flutter-package/issues/2)

## [0.0.6]- 09 May 2020

- added background property for textfield to create borderLess Passwordfield with padding for the text
- Improved documentation and updated example

## [0.0.5]- 11 Jan 2020

- passwordfield now supports adding constraints on password in the form of regex pattern which is fully customisable.
- deprecated onchanged method and can no longer be used.
- minor improvements and changes.

## [0.0.4]- 06 Sep 2019

- added focusedBorder and border for the passwordField that allows more fexibility to decorate the passwordfield's border- color,width,radius & style.

## [0.0.3]- 03 Sep 2019

- minor changes and added support for border for the passwordfield

## [0.0.1]- 25 Aug 2019

- Initial Release of the Package with basic functionality.

## [0.0.2]- 26 Aug 2019

- updated readme file with examples and explanation


test/passwordfield_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:passwordfield/src/passwordfield.dart';

void main() {
  Widget _boilerplate({required Widget child}) {
    return MaterialApp(
      home: Directionality(
        textDirection: TextDirection.ltr,
        child: MediaQuery(
          data: const MediaQueryData(size: Size(800.0, 600.0)),
          child: Center(
            key: const Key('centerKey'),
            child: Material(
              child: child,
            ),
          ),
        ),
      ),
    );
  }

  Finder _findText(String text) {
    return find.text(text);
  }

  const String errorMessage = 'required at least 1 letter and number 5+ chars';
  const String invalidString = 'password';
  const String validString = 'password123';
  const String regexpattern = r'^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{5,}$';
  group('Passwordfield integration test', () {
    testWidgets('passwordfield loads', (WidgetTester tester) async {
      await tester.pumpWidget(_boilerplate(
          child: PasswordField(
        key: const Key('passwordFieldKey'),
      )));
      await tester.pump();
      expect(find.byKey(const Key('passwordFieldKey')), findsOneWidget);
    });

    testWidgets('sets initial text', (WidgetTester tester) async {
      await tester.pumpWidget(_boilerplate(
          child: PasswordField(
        key: const Key('passwordFieldKey'),
        controller: TextEditingController(text: 'password'),
      )));
      await tester.pump();
      expect(find.byKey(const Key('passwordFieldKey')), findsOneWidget);
      expect(_findText('password'), findsOneWidget);
    });

    testWidgets('Widget loads with no error', (WidgetTester tester) async {
      await tester.pumpWidget(_boilerplate(
        child: PasswordField(
          key: const Key('passwordFieldKey'),
          errorMessage: errorMessage,
          passwordConstraint: regexpattern,
          passwordDecoration: PasswordDecoration(
            inputStyle: const TextStyle(
              fontSize: 14,
            ),
          ),
          hintText: 'Password',
        ),
      ));

      await tester.pumpAndSettle();
      final field = find.byKey(const Key('passwordFieldKey'));
      final errorNotFound = _findText(errorMessage);
      await tester.pump();
      expect(field, findsOneWidget);
      expect(errorNotFound, findsNothing);
    });
  });

  testWidgets('Enter a invalid string throws error',
      (WidgetTester tester) async {
    await tester.pumpWidget(_boilerplate(
      child: PasswordField(
        key: const Key('passwordFieldKey'),
        errorMessage: errorMessage,
        passwordConstraint: regexpattern,
        passwordDecoration: PasswordDecoration(
          inputStyle: const TextStyle(
            fontSize: 14,
          ),
        ),
        hintText: 'Password',
      ),
    ));

    final field = find.byKey(const Key('passwordFieldKey'));

    await tester.enterText(field, invalidString);
    final invalidText = _findText(invalidString);
    final findsError = _findText(errorMessage);
    await tester.pump();
    expect(invalidText, findsOneWidget);
    expect(findsError, findsOneWidget);
  });

  testWidgets('Valid String hides the error', (WidgetTester tester) async {
    await tester.pumpWidget(_boilerplate(
      child: PasswordField(
        key: const Key('passwordFieldKey'),
        errorMessage: errorMessage,
        passwordConstraint: regexpattern,
        passwordDecoration: PasswordDecoration(
          inputStyle: const TextStyle(
            fontSize: 14,
          ),
        ),
        hintText: 'Password',
      ),
    ));

    final field = find.byKey(const Key('passwordFieldKey'));
    await tester.enterText(field, validString);
    final validText = _findText(validString);
    final errorFinder = _findText(errorMessage);
    await tester.pump();
    expect(validText, findsOneWidget);
    expect(errorFinder, findsNothing);
  });

  testWidgets('floatingPlaceholder must have a floating text',
      (WidgetTester tester) async {
    Widget passwordWidget() => PasswordField(
          key: const Key('passwordFieldKey'),
          errorMessage: errorMessage,
          passwordConstraint: regexpattern,
          passwordDecoration: PasswordDecoration(
            inputStyle: const TextStyle(
              fontSize: 14,
            ),
          ),
          hintText: 'Password',
          floatingText: '',
          hasFloatingPlaceholder: true,
        );

    await tester.pump();
    expect(() => passwordWidget(), throwsAssertionError);
  });

  testWidgets('Must satisfy max length constraint',
      (WidgetTester tester) async {
    final controller = TextEditingController();
    const int maxLength = 10;

    Widget passwordWidget() => PasswordField(
          key: const Key('passwordFieldKey'),
          errorMessage: errorMessage,
          maxLength: maxLength,
          passwordConstraint: regexpattern,
          controller: controller,
          passwordDecoration: PasswordDecoration(
            inputStyle: const TextStyle(
              fontSize: 14,
            ),
          ),
          hintText: 'Password',
        );

    await tester.pumpWidget(_boilerplate(
      child: passwordWidget(),
    ));

    final field = find.byKey(const Key('passwordFieldKey'));
    await tester.enterText(field, 'morethan10characters');

    /// greater than 10 characters
    await tester.pumpAndSettle();
    expect(controller.text.length, maxLength);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';

import 'package:example/main.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/src/password_bloc.dart
import 'dart:async';

class PasswordBloc extends Validation {
  final _passwordController = StreamController<List<String>>.broadcast();

// validating the password
  Stream<String> get password =>
      _passwordController.stream.transform(validatePassword);

  void onPasswordChanged(String pattern, String text) {
    final List<String> list = [pattern, text];
    _passwordController.sink.add(list);
  }

  void dispose() {
    _passwordController.close();
  }
}

class Validation {
  var validatePassword = StreamTransformer<List<String>, String>.fromHandlers(
      handleData: (list, sink) {
    final String pattern = list[0];
    final String password = list[1];
    final RegExp regex = RegExp('''$pattern''');
    const int minLength = 8;
    final bool hasUppercase = password.contains(RegExp(r'[A-Z]'));
    final bool hasDigits = password.contains(RegExp(r'[0-9]'));
    final bool hasLowercase = password.contains(RegExp(r'[a-z]'));
    final bool hasSpecialCharacters =
        password.contains(RegExp(r'[!@#$%^&*(),.?":{}|<>]'));
    final bool hasMinLength = password.length > minLength;

    if (pattern.isEmpty) {
      if (hasDigits &
          hasUppercase &
          hasLowercase &
          hasSpecialCharacters &
          hasMinLength) {
        sink.add(password);
      } else {
        sink.addError('Invalid password format :(');
      }
    } else {
      if (regex.hasMatch(password)) {
        sink.add(password);
      } else {
        sink.addError('Invalid password format :(');
      }
    }
  });
}


lib/src/passwordfield.dart
library passwordfield;

import 'package:flutter/material.dart';
import 'package:passwordfield/src/password_bloc.dart';

class PasswordField extends StatefulWidget {
  PasswordField({
    Key? key,
    this.autoFocus = false,
    this.backgroundColor,
    this.border,
    this.controller,
    this.color,
    this.errorMaxLines,
    this.errorMessage,
    this.floatingText = 'Password',
    this.hasFloatingPlaceholder = false,
    this.hintText = 'Password',
    this.passwordDecoration,
    this.maxLength,
    this.onSubmit,
    this.onChanged,
    this.passwordConstraint = '',
  })  : assert(
            (hasFloatingPlaceholder == true && floatingText!.isNotEmpty) ||
                hasFloatingPlaceholder == false,
            'FloatingText cannot be null or empty when hasFloatingPlaceholder is true'),
        super(key: key);

  /// if autofocus is true keyboard pops up as soon as the widget is rendered on screen
  /// defaults to false
  final bool? autoFocus;

  /// Background Color for the textfield (defaults to white)
  /// By default the backgroundColor is based on the current [Theme].
  final Color? backgroundColor;

  /// Input Border for the passwordfield
  final PasswordBorder? border;

  /// A controller for an editable passwordfield.
  final TextEditingController? controller;

  /// changes the primary color of the PasswordField
  final Color? color;

  /// The maximum number of lines the [errorMessage] can occupy.
  ///
  /// Defaults to null, which means that the [errorMessage] will be limited
  /// to a single line with [TextOverflow.ellipsis].
  final int? errorMaxLines;

  /// custom message to show if the input password does not match the pattern.
  final String? errorMessage;

  /// if hasFloatingPlaceholder==true
  /// a text label floats to left top on focus
  /// The label defaults to "Password" if not specified,
  ///
  /// floating text can be styled using [PasswordDecoration.hintStyle]
  ///
  /// Note: either [floatingText]/ [hintText] can be shown at a time
  /// that mainly depends on property [hasFloatingPlaceholder]
  final String? floatingText;

  /// whether the placeholder can float to left top on focus, defaults to false
  final bool hasFloatingPlaceholder;

  ///default text to show on the passwordfield when it is empty
  final String? hintText;

  /// the max number of characters the password field can support
  final int? maxLength;

  /// function triggerred when the submit button on keyboard is pressed
  final Function(String)? onSubmit;

  /// A Callback function triggered when the text inside the PasswordField changes
  ///
  final Function(String)? onChanged;

  /// decoration for the input
  final PasswordDecoration? passwordDecoration;

  ///
  /// RegEx pattern for the input password
  ///
  ///     r'[a-zA-Z]'      // 'heLLo' allows Alphabets with upper and lower case
  ///     r'[a-zA-Z]{8}'   // 'helloYou' allows Alphabetic password strict to 8 chars
  ///     r'[0-9a-zA-Z]';  // 'Hello123' allows alphanumeric password
  ///     r'[0-9]{6}'      //  '123456' allows numeric password strict to 6 characters
  ///
  /// if the pattern is not specified then by default the password must satisfy
  /// - A uppercase letter
  /// - A lowercase letter
  /// - A digit
  /// - A special character
  /// - A minimum length of 8 characters
  /// Dart regular expressions have the same syntax and semantics
  /// as JavaScript regular expressions.
  /// See:[ecma-international.org/ecma-262/9.0/#sec-regexp-regular-expression-objects](ecma-international.org/ecma-262/9.0/#sec-regexp-regular-expression-objects)
  /// for the specification of JavaScript regular expressions.
  ///
  final String passwordConstraint;

  @override
  State createState() => PasswordFieldState();
}

class PasswordFieldState extends State<PasswordField> {
  bool obscureText = true;
//wrap your toggle icon in Gesture Detector

  void inContact(TapDownDetails details) {
    setState(() {
      obscureText = false;
    });
  }

  void outContact(TapUpDetails details) {
    setState(() {
      obscureText = true;
    });
  }

  PasswordBloc bloc = PasswordBloc();
  final underlineBorder = const UnderlineInputBorder();

  @override
  Widget build(BuildContext context) {
    final defaultTextStyle = DefaultTextStyle.of(context).style;
    widget.passwordDecoration?.inputStyle ??= defaultTextStyle;
    return Theme(
      data: ThemeData(
        primaryColor: widget.color ?? Theme.of(context).primaryColor,
        colorScheme: Theme.of(context)
            .colorScheme
            .copyWith(primary: widget.color ?? Theme.of(context).primaryColor),
      ),
      child: StreamBuilder<String>(
        stream: bloc.password,
        builder: (BuildContext context, AsyncSnapshot snapshot) {
          InputDecoration _buildDefaultInputDecoration() {
            return InputDecoration(
                fillColor: widget.backgroundColor,
                filled: widget.backgroundColor != null,
                suffixIcon: GestureDetector(
                  child: const Icon(Icons.remove_red_eye),
                  onTapDown: inContact,
                  onTapUp: outContact,
                ),
                hintText: widget.hintText,
                errorText: snapshot.hasError
                    ? widget.errorMessage ?? snapshot.error as String?
                    : null,
                errorMaxLines: widget.errorMaxLines,
                border: widget.border == null
                    ? underlineBorder
                    : widget.border?.border,
                enabledBorder: widget.border == null
                    ? underlineBorder
                    : widget.border?.enabledBorder,
                focusedBorder: widget.border == null
                    ? underlineBorder
                    : widget.border?.focusedBorder,
                focusedErrorBorder: widget.border == null
                    ? null
                    : widget.border?.focusedErrorBorder,
                errorBorder:
                    widget.border == null ? null : widget.border!.errorBorder,
                floatingLabelBehavior: widget.hasFloatingPlaceholder
                    ? FloatingLabelBehavior.auto
                    : FloatingLabelBehavior.never);
          }

          return TextField(
              maxLength: widget.maxLength,
              controller: widget.controller,
              obscureText: obscureText,
              autofocus: widget.autoFocus!,
              decoration: widget.passwordDecoration == null
                  ? _buildDefaultInputDecoration()
                  : widget.passwordDecoration!.copyWith(
                      fillColor: widget.backgroundColor,
                      filled: widget.backgroundColor != null,
                      contentPadding: widget.passwordDecoration!.inputPadding,
                      errorText: snapshot.hasError
                          ? widget.errorMessage ?? snapshot.error as String?
                          : null,
                      errorMaxLines: widget.errorMaxLines,
                      errorStyle: widget.passwordDecoration!.errorStyle ??
                          defaultTextStyle.copyWith(color: Colors.red),
                      hintText: widget.hintText,
                      hintStyle: widget.passwordDecoration?.hintStyle ??
                          widget.passwordDecoration?.inputStyle,
                      labelStyle: widget.passwordDecoration!.hintStyle ??
                          widget.passwordDecoration!.inputStyle,
                      border: widget.border == null
                          ? underlineBorder
                          : widget.border?.border,
                      enabledBorder: widget.border == null
                          ? underlineBorder
                          : widget.border?.enabledBorder,
                      focusedBorder: widget.border == null
                          ? underlineBorder
                          : widget.border?.focusedBorder,
                      focusedErrorBorder: widget.border == null
                          ? null
                          : widget.border?.focusedErrorBorder,
                      errorBorder: widget.border == null
                          ? null
                          : widget.border!.errorBorder,
                      counterText: '',
                      floatingLabelBehavior: widget.hasFloatingPlaceholder
                          ? FloatingLabelBehavior.auto
                          : FloatingLabelBehavior.never,
                      labelText: widget.hasFloatingPlaceholder
                          ? widget.floatingText ?? widget.hintText
                          : widget.hintText,
                      suffixIcon: GestureDetector(
                        child: widget.passwordDecoration!.suffixIcon,
                        onTapDown: inContact,
                        onTapUp: outContact,
                      )),
              onSubmitted: widget.onSubmit,
              style: widget.passwordDecoration?.inputStyle,
              onChanged: (text) {
                bloc.onPasswordChanged(widget.passwordConstraint, text);
                if (widget.onChanged != null) {
                  widget.onChanged!(text);
                }
              });
        },
      ),
    );
  }
}

/// Decoration class for the PasswordField to customize the input styling
class PasswordDecoration extends InputDecoration {
  PasswordDecoration(
      {this.hintStyle,
      this.inputStyle,
      this.errorStyle,
      this.inputPadding,
      this.suffixIcon = const Icon(Icons.remove_red_eye)});

  /// styling fpr the the hint and the floating label,
  /// defaults to same as inputStyle if not specified
  final TextStyle? hintStyle;

  /// styling the Passwordfield Text
  TextStyle? inputStyle;

  /// style for the the errorMessage
  final TextStyle? errorStyle;

  /// Input padding
  final EdgeInsetsGeometry? inputPadding;

  /// Icon used to unhide the password when touch in Contact with the icon
  final Icon? suffixIcon;
}

/// Consolidated Border class for the passwordfield
/// when not in focus
class PasswordBorder {
  PasswordBorder(
      {this.border,
      this.focusedBorder,
      this.enabledBorder,
      this.focusedErrorBorder});

  /// Input Border for the passwordfield when not in focus.
  InputBorder? border;

  ///  Input Border for the passwordfield when field is not disabled.
  InputBorder? enabledBorder;

  ///  Input Border for the passwordfield when field is unfocused and has error.
  InputBorder? errorBorder;

  ///  Input Border for the passwordfield when field is focused and has error.
  InputBorder? focusedErrorBorder;

  ///  Input Border for the passwordfield when in focus.
  InputBorder? focusedBorder;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:passwordfield/passwordfield.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
          primaryColor: Colors.red,
          colorScheme:
              ThemeData.light().colorScheme.copyWith(primary: Colors.red)),
      home: PasswordApp(),
    );
  }
}

class PasswordApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        FocusScope.of(context).unfocus();
      },
      child: Scaffold(
          appBar: AppBar(
            title: const Text('PasswordField Demo'),
          ),
          body: ListView(
            children: <Widget>[
              const SizedBox(
                height: 20,
              ),
              Padding(
                padding:
                    const EdgeInsets.symmetric(horizontal: 20, vertical: 20),
                child: PasswordField(
                  backgroundColor: Colors.red.shade100,
                  controller: TextEditingController(text: 'password'),
                  errorMessage:
                      'required at least 1 letter and number 5+ chars',
                  passwordConstraint: r'^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{5,}$',
                  passwordDecoration: PasswordDecoration(
                    inputPadding: const EdgeInsets.symmetric(horizontal: 10),
                    suffixIcon: const Icon(
                      Icons.not_accessible,
                      color: Colors.grey,
                    ),
                    inputStyle: const TextStyle(
                      fontSize: 14,
                    ),
                  ),
                  hintText: 'Password',
                  onChanged: (x) {
                    print(x);
                  },
                  border: PasswordBorder(
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(12),
                    ),
                    focusedBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(12),
                    ),
                    focusedErrorBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(12),
                      borderSide:
                          BorderSide(width: 2, color: Colors.red.shade200),
                    ),
                  ),
                ),
              ),
              const SizedBox(
                height: 20,
              ),
              Padding(
                padding: const EdgeInsets.symmetric(horizontal: 16),
                child: PasswordField(
                  color: Colors.blue,
                  passwordConstraint: r'.*[@$#.*].*',
                  passwordDecoration: PasswordDecoration(),
                  // hintText: 'must have special characters',
                  border: PasswordBorder(
                    border: OutlineInputBorder(
                      borderSide: BorderSide(
                        color: Colors.blue.shade100,
                      ),
                      borderRadius: BorderRadius.circular(12),
                    ),
                    focusedBorder: OutlineInputBorder(
                      borderSide: BorderSide(
                        color: Colors.blue.shade100,
                      ),
                      borderRadius: BorderRadius.circular(12),
                    ),
                    focusedErrorBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(12),
                      borderSide:
                          BorderSide(width: 2, color: Colors.red.shade200),
                    ),
                  ),
                  errorMessage:
                      'must contain special character either . * @ # \$',
                ),
              ),
              const SizedBox(
                height: 20,
              ),
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: PasswordField(
                  backgroundColor: Colors.blue.withOpacity(0.2),
                  errorMessage: '''
- A uppercase letter
- A lowercase letter
- A digit
- A special character
- A minimum length of 8 characters
                 ''',
                  hintText: 'Default password constraint ',
                  passwordDecoration: PasswordDecoration(
                    inputPadding: const EdgeInsets.symmetric(horizontal: 20),
                  ),
                  border: PasswordBorder(
                    border: OutlineInputBorder(
                        borderSide:
                            const BorderSide(width: 0, color: Colors.grey),
                        borderRadius: BorderRadius.circular(25.7)),
                    focusedBorder: OutlineInputBorder(
                      borderSide:
                          const BorderSide(width: 0, color: Colors.grey),
                      borderRadius: BorderRadius.circular(25.7),
                    ),
                    enabledBorder: OutlineInputBorder(
                      borderSide:
                          const BorderSide(width: 0, color: Colors.grey),
                      borderRadius: BorderRadius.circular(25.7),
                    ),
                  ),
                ),
              ),
            ],
          )),
    );
  }
}
